### PR TITLE
[READY] The Portal Update - Linked portals, momentum conservation, and more!

### PR DESCRIPTION
**Pull request by silicons**
- Portals are now non dense, and invalid for using as spacemove backups
- Anything that crosses a portal will be teleported with no inaccuracy
- Inaccuracy for hand teleporter portals will be calculated on portal opening
- Hand teleporters, mech portals, and portal guns now use LINKED PORTALS
- When a linked portal is open, a matching portal is opened at the destination
- Anything crossing either will be moved to the other.
- You can use a hand teleporter to close its own portals
- Projectiles and thrown objects will pass through the portals with full momentum conserved.
- Normal machinery teleporters, quantum pads, wormhole jaunters, and spacetime anomaly event wormholes are unaffected and will still be one-way.
- WIP: Allowing people to see through linked portals!

Pictures to come soon!

For people who don't frequent VGstation github, it's obviously inspired by https://github.com/vgstation-coders/vgstation13/pull/6032

### CODE REVIEW

**Comment by Razharas** [comment]
>Portals are now non dense, and invalid for using as spacemove backups

This should change to reflect the properties of whats on the other side, if theres a grille right by the other portal you can use it as it if was in place on the space one
Same with density actually, if the other portal leads into a wall or other completely impassable object you should just bump that object from the place of the original portal and stay in place

**Comment by GunHog** [comment]
This is neato, Kev! I cannot wait to **test** this! If I could add some creep to your feature, could we have stationary portals, too?

**Comment by Razharas** [comment]
Oh and you can steal rrs zlevel shadow code and repurpose it to make a honest to god portals with actual visuals of destination inside the coloured circle

**Comment by Razharas** [comment]
And you can add the portal-connected turfs to eachother-s connected turfs list and have a real easy atmos connectivity through portals(this is getting really fancy)

**Comment by kevinz000** [comment]
@Razharas 
For atmos connectivity I left it out for balance reasons, if people think it's a good idea I or someone else could PR it at a later date though.
As for .. zlevel.. shadow... code... wot... mind.. blown yeah i'll ask remie later
as for spacemove backup and density stuff yeah i'll get to it
@GunHog lol

File: code/game/objects/effects/portals.dm
Comment (RemieRichards) [inline code/game/objects/effects/portals.dm]: this should be ```P1.link_portal(P2)``` surely?, for consistency.

**Comment by RemieRichards** [comment]
@Razharas nah, visuals for portals should wait till 512, it can probably use ```visual_contents```.

File: code/game/objects/effects/portals.dm
Comment (kevinz000) [inline code/game/objects/effects/portals.dm]: woops

**Comment by PKPenguin321** [comment]
Very cool and I want it but I also want to take bets on how many new issues will be opened once this is merged, guessing around 15

File: code/game/objects/effects/portals.dm
Comment (Razharas) [inline code/game/objects/effects/portals.dm]: In the link_portal you call this before you cann this proc without any checks, i dont know why you check for anything here and try to unlink them yet again
I dont know why you unlink them at all to be honest

File: code/game/objects/effects/portals.dm
Comment (Razharas) [inline code/game/objects/effects/portals.dm]: the comment you deleted above was very true

File: code/game/objects/effects/portals.dm
Comment (Razharas) [inline code/game/objects/effects/portals.dm]: This is redundant, by nature of how its done you cant accidentally link lists twice, its a binary state by the design(assoc lists of turfs to true or false statements)

File: code/game/objects/effects/portals.dm
Comment (Razharas) [inline code/game/objects/effects/portals.dm]: This proc doesnt check if the turfs are adjucent, you can unlink nearby turfs from one another like this and they will stay unlinked

File: code/game/objects/effects/portals.dm
Comment (Razharas) [inline code/game/objects/effects/portals.dm]: This is doubly redundant since you already call this in the proc that calls this one, you also check for some shit for some realon while there you just call it without care
Remove it all

**Review by Razharas — DISMISSED** [review]

**Comment by Grazyn** [comment]
Static portals seem like something too OP even for science standards

**Comment by Tacolizard** [comment]
@Grazyn what do you mean by 'static portals' afaik this pr isn't adding anything new, just improving.

File: code/game/objects/effects/portals.dm
Comment (Cyberboss) [inline code/game/objects/effects/portals.dm]: remove the `= null`s

File: code/game/objects/effects/portals.dm
Comment (Cyberboss) [inline code/game/objects/effects/portals.dm]: . = INITIALIZE_HINT_QDEL
CRASH("Somebody fucked up")

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by kevinz000** [comment]
@RemieRichards I needed a way to check ifs omething is coming from another portal and I don't have a better method, because otherwise it's just going to infinitely teleport the mob around both..

File: code/game/atoms_movable.dm
Comment (kevinz000) [inline code/game/atoms_movable.dm]: Um this is kinda necessary for tracking if someone already passed through a portal...

**Comment by fludd12** [comment]
For seeing through the portal, would it just show the ground, or take your perspective and let you see around it?

Also, will atmos and speech go through so you can replicate the last scene of Portal 2?

**Comment by kevinz000** [comment]
@fludd12 
1. not implemented, will do later to not hold up pr
2. Disabled for balance reasons
3. not implemented, will do later to not hold up PR

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/game/atoms_movable.dm
Comment (Cyberboss) [inline code/game/atoms_movable.dm]: Remove this

File: code/game/objects/effects/portals.dm
Comment (Cyberboss) [inline code/game/objects/effects/portals.dm]: name the type var something other than a byond keyword please

File: code/game/objects/effects/portals.dm
Comment (Cyberboss) [inline code/game/objects/effects/portals.dm]: wouldn't forcemoving to the portals turf already trigger a teleport?

File: code/game/objects/effects/portals.dm
Comment (Cyberboss) [inline code/game/objects/effects/portals.dm]: wouldn't forcemoving to the portals turf already trigger a teleport?

File: code/game/objects/effects/portals.dm
Comment (Cyberboss) [inline code/game/objects/effects/portals.dm]: What @razharas said

File: code/modules/projectiles/guns/energy/special.dm
Comment (Cyberboss) [inline code/modules/projectiles/guns/energy/special.dm]: else if

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/game/atoms_movable.dm
Comment (Cyberboss) [inline code/game/atoms_movable.dm]: document!

**Review by Cyberboss — DISMISSED** [review]

File: code/game/objects/effects/portals.dm
Comment (kevinz000) [inline code/game/objects/effects/portals.dm]: again this is a safety check so incase they're linked to two different locations somehow you don't have random wormholes of air happening. are we absolutely sure this should be removed??

Comment (kevinz000) [inline code/game/objects/effects/portals.dm]: link_atmos() isn't even called much, only on creation (unless someone adds mobile portals) so this isn't exactly a costly check.

**Comment by adamkad1** [comment]
portal 13?

**Comment by Cyberboss** [comment]
If this breaks I'm revoking your freeze bypass

**Pull request merged**
